### DISCUSSION TITLE
Ability to customise API middleware error responses

### DISCUSSION THREAD

**Discussion by satazor** [vercel/next.js#9957] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

Next.js has API routes with built-in middlewares, such as `bodyParser`. When an error is thrown by those middlewares, Next.js sends an error response, handled by [`sendError`](https://github.com/zeit/next.js/blob/00179adcaa3140f74786d4cd07c0adc7ad6ae01d/packages/next/next-server/server/api-utils.ts#L251).

The current implementation sends a JSON response with the error message as the body. As an example: if we send request, with `Content-Type: application/json` with an invalid JSON request body, the server responds with `Invalid JSON`.

However, usually APIs that I develop have a standard envelope for errors, like:

```json
{
  "statusCode": 400,
  "message": "Invalid request body",
  "other": "optional stuff here"
}
```

I would like to be able to keep the same response envelope for built-in middleware errors.

## Describe the solution you'd like

Perhaps, we could somehow be able to tweak the `sendError` by including it in the `config` exported in API routes?

```js
export const config = {
  api: {
    sendError: (res, statusCode, message) => {},
  },
};
```

## Describe alternatives you've considered

I've looked into the code to see if I could somehow monkey-patch to get around the issue temporarily, but I couldn't find a way. Another solution is to turn off all the middleware and use something like micro to have full-control, but seems overkill.

## Additional context

<img width="358" alt="Screenshot 2020-01-05 at 22 06 09" src="https://user-images.githubusercontent.com/1017236/71786781-96bb8f00-3007-11ea-8e16-b7fbe98867b1.png">

**Comment by fastfrwrd** [comment]
maybe a `pages/api/_error.ts` could be good, it would feel similar to error handling on the React side of things.

**Comment by satazor** [comment]
> maybe a `pages/api/_error.ts` could be good, it would feel similar to error handling on the React side of things.

That looks like a great idea actually!
